### PR TITLE
Feature/delete api dfu target

### PR DESCRIPTION
**Pull request by sigvartmh**
Not complete yet but my idea of how the API is supposed to be is there.

### CODE REVIEW

File: include/dfu/dfu_target.h [hunk 1]
```diff
@@ -36,6 +36,7 @@ struct dfu_target {
 	int (*init)(size_t file_size, dfu_target_callback_t cb);
 	int (*offset_get)(size_t *offset);
 	int (*write)(const void *const buf, size_t len);
+	int (*erase)(void);
```
Comment (lemrey) [inline include/dfu/dfu_target.h:39]: I would let it be asynchronous and take a callback instead of block. Erasing memory takes a long time.

Comment (sigvartmh) [reply include/dfu/dfu_target.h:39]: Agreed that's probably best so that people can decide to make it blocking or not.

**Comment by tokangas** [comment]
Thanks @sigvartmh. The current implementation in dfu_target_modem.c seems to be such, that already dfu_target_init() erases the modem backup. There's a check in dfu_target_modem_init() that if the offset returned by modem is 0x280000 (which it is after an update), dfu_target_modem_erase() is called. So it seems that calling dfu_target_erase() is actually just trying to do the same thing again. Also calling dfu_target_reset() after erasing results in "Modem upgrade aborted." trace from dfu_target_modem_done(), which is a bit misleading.

Have you though about how revert could be implemented? Just adding a similar function for revert doesn't work, because the backup would always be erased when dfu_target_init() is called, so revert would never be possible.

**Comment by sigvartmh** [comment]
> Thanks @sigvartmh. The current implementation in dfu_target_modem.c seems to be such, that already dfu_target_init() erases the modem backup. There's a check in dfu_target_modem_init() that if the offset returned by modem is 0x280000 (which it is after an update), dfu_target_modem_erase() is called. So it seems that calling dfu_target_erase() is actually just trying to do the same thing again. Also calling dfu_target_reset() after erasing results in "Modem upgrade aborted." trace from dfu_target_modem_done(), which is a bit misleading.

Okey I see there are some issues with this approach. I just don't know how I can make it generic without forcing people to call erase after dfu_init(maybe that's the solution). I thought to keep some of the sanity checks like if the offset is dirty delete etc. would be beneficial but seems they are actually causing the issue.

I'll try to clean up the error messages in the log and see if I can figure out a way to get it to work.

> 
> Have you though about how revert could be implemented? Just adding a similar function for revert doesn't work, because the backup would always be erased when dfu_target_init() is called, so revert would never be possible.

I think due to this condition erase should be removed from init. So that revert will be possible thanks for spotting it had not thought about revert when implementing this. I'll change it so that it fits better what is needed.

**Comment by tokangas** [comment]
> I think due to this condition erase should be removed from init. So that revert will be possible thanks for spotting it had not thought about revert when implementing this. I'll change it so that it fits better what is needed.

Sounds good, it might be a good idea that erasing doesn't happen automatically, but needs to be explicitly called.

**Pull request closed without merging**